treeherder/services/pulse/connection.py

```diff
@@ -1,18 +1,19 @@
 import environ
+from django.utils.functional import lazy
 from kombu import Connection
 
 env = environ.Env()
 
-# Used to specify the PulseGuardian account that will be used to create
-# ingestion queues for the exchanges specified in ``PULSE_SOURCE_EXCHANGES``.
-# See https://pulse.mozilla.org/whats_pulse for more info.
-# Example: "amqp://myuserid:mypassword@pulse.mozilla.org:5672/?ssl=1"
-config = env.url("PULSE_URL")
 
+def build_connection():
+    """Build a Kombu Broker connection to Mozilla Pulse."""
+    # Used to specify the PulseGuardian account that will be used to create
+    # ingestion queues for the exchanges specified in ``PULSE_SOURCE_EXCHANGES``.
+    # See https://pulse.mozilla.org/whats_pulse for more info.
+    # Example: "amqp://myuserid:mypassword@pulse.mozilla.org:5672/?ssl=1"
+    pulse_url = env.url("PULSE_URL")
 
-def build_connection(url):
-    """Build a Kombu Broker connection to Mozilla Pulse with the given url."""
-    return Connection(url)
+    return Connection(pulse_url.geturl())
 
 
-pulse_conn = build_connection(config.geturl())
+pulse_conn = lazy(build_connection, Connection)()
```

Comment:
The docstring for `lazy` says "Results are not memoized; the function is evaluated on every access".

Does this mean the Pulse connection will be recreated at every access? (ie defeating connection pooling)

Reply:
It depends on how the connection is used.  If the user imports, calls and then discards the result then a new connection would be initialised when they next call it again.  I've used it as a context manager in the `read_pulse_*` commands so the connection is persistent for the lifetime of each script.  They were [originally](https://github.com/mozilla/treeherder/blob/4c565943df7c1fd5b9ddc967afa0c4ce6359d320/treeherder/etl/management/commands/read_pulse_jobs.py#L30) constructing a connection in a similar fashion.

We could  enhance the object further to make this a lazy singleton connection but that should be another PR so we don't block all the other changes (and because I'm not entirely sure how I'd tackle that without some testing!).

Reply:
Ok sounds good :-)